### DISCUSSION TITLE
Translate button on comments

### DISCUSSION THREAD

**Discussion by Cyan903** [Docile-Alligator/Infinity-For-Reddit#970] [Ideas]
A button to translate comments on posts would be cool. Similar to how YouTube and Twitter have their translate buttons, except maybe it could be placed next to the "see removed comment" button as it would get in the way of the normal UI.

| ![image](https://user-images.githubusercontent.com/88910588/159147358-ce0f5e4e-0653-4539-b71d-68fd7fc98e88.png) | ![image](https://user-images.githubusercontent.com/88910588/159147375-8d931daf-8ae4-4a63-a2e1-686ae940fcb5.png) |
|-----------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------|

I know this would have to rely on an API of some sort but I wouldn't mind if I had to supply my own key within the app or something. I think it would be a nice feature especially on multilingual subreddits.

**Comment by Generator** [comment]
Lingva could provide privacy friendly API (and multiple instances)  
https://github.com/TheDavidDelta/lingva-translate#public-apis

In alternative there's Deepl  
https://www.deepl.com/docs-api

**Comment by Wladefant** [comment]
Duplicate: #853

**Comment by Wladefant** [comment]
This is a very good function, that should be implemented, i really like the translate function in the youtube comments and would therefore recommend to implement something similar. This is an example in Youtube regarding the frontend:
![image](https://user-images.githubusercontent.com/79611093/178480418-e1931863-f3ed-4dc6-9501-031a1964e97f.png)

When we speak about the backend I would suggest we could use the Translation ML Kit from Google. This is an on-device Translation and supports more than 50 languages. These are offline models, that are also used in google translate offline mode, and therefore not as good as the google online translation.
If we want to make the implementation easy and intuitive, we could use the solution mentioned above and just follow this tutorial: https://developers.google.com/ml-kit/language/translation

Something worth mentioning is, that the models are all trained for translation from and to the English language, so the English language will always be used for translation.
Another thing is that we would need a language identification tool to make the translation enjoyable.

If anyone is interested in this feature please consider upvoting and let your comments, because there are also other solutions that i could present if other people are interested. 
@Docile-Alligator what do you think?

**Reply by skal-dB** [reply]
(my two cents) 

At a personal level I would love to see DeepL running together with Infinity. 
It make sense with the open source side of the force, and it works pretty fine. Even with Dutch (wich is a solid achievement).

Also the app can recognize the original language, as GTranslate does.

**Comment by filex100** [comment]
I'd like to have this feature for all content, but it wouldn't be necessary to use an external API (or so I think) since Reddit already does some translations. 

They won't be perfect, but they would help occasionally. 

Could this be considered again?

**Comment by user-8tw6i** [comment]
This would be a nice feature, any updates on this?

I also like the idea of having an option to provide API Key, something similar provides [TranslateYou](https://github.com/you-apps/TranslateYou) where user can choose free / paid API (there are 15 pre-defined options).

**Comment by sanjuant** [comment]
**Local-first translate option (Bergamot/WebAssembly) + pluggable providers**

Love this idea and +1 to the privacy-friendly angle. In addition to cloud APIs (DeepL, Lingva, etc.) or Reddit’s own translations, could we add a **local/on-device translation** option similar to Firefox Translator?
Reference: [https://github.com/DavidVentura/firefox-translator](https://github.com/DavidVentura/firefox-translator) (Bergamot models running in WASM).

**Why local-first?**

* **Privacy**: no comment text leaves the device.
* **Offline**: works even with flaky connections.
* **Cost control**: zero API calls by default; users can still plug in their own API keys for top-tier quality.

**How it could look (high level):**

* **Provider interface** with three families:

  1. **Local** (Bergamot via WASM, like firefox-translator)
  2. **Built-in/Reddit** (when available for a comment)
  3. **External APIs** (user-supplied keys: DeepL, Google, etc.)
* **UI**: a “Translate” button next to the comment → shows the translated text inline. Add a small chevron/dropdown to pick target language and provider (Local / Reddit / API-X). Remember the user’s last choice.
* **Language detection**: include a lightweight detector (e.g., whatlang/franc/CLD3) to auto-pick source and avoid unnecessary model downloads.
* **Model management**: download per language (or language→EN) on first use; cache and show size before download. Bergamot typically pivots through English; that aligns with many systems trained to/from EN.
* **Fallback policy**: if the local model isn’t available or fails, optionally fall back to Reddit/API (configurable: strict local-only vs. allow fallback).
* **Caching**: memoize translations per comment ID + hash to avoid reprocessing.

**Notes / trade-offs:**

* Local models are larger and often a bit lower quality than top cloud MT, but they’re plenty good for “understand this comment” use cases.
* Let users opt into specific language packs to manage disk usage.
* Keep the provider layer simple so the community can add more backends later.

If this direction sounds useful, I support adding it and I’m happy to share usage feedback once it’s available.
